### PR TITLE
[PERF] Join FWD + BWD and Sync Tensors Precisely

### PR DESCRIPTION
**Pull request by pglusacTT**
### Issue
N/A

---

### Problem description

By default, tt-torch backend inside tt-xla does some variant of `torch_xla.sync()` [call during model execution](https://github.com/tenstorrent/tt-xla/blob/a0ba4ca3266475f441f3d326d612d5659c5845ae/python_package/tt_torch/backend/backend.py#L272). 
[Recent PR](https://github.com/tenstorrent/tt-xla/pull/5235) on the tt-xla repo allows for purely lazy execution. 

This allows us to lazily join FWD & BWD graphs and to be surgical about which tensors we should sync. 

This yields performance improvements since we avoid redundant computation.

---

### Perf Improvement noticed on Mixa's QB2:

**LLaMA 3.1 8B, BS=1 SEQ=512**

| Old | New | Improvement |
| --- | --- | --- |
5.13s | 3.74s | 27.2%

**LLaMA 70B L=40, BS=1 SEQ=512**
| Old | New | Improvement |
| --- | --- | --- |
2.156s | 1.858s | 13.82%

Training curve and loss values don't seem to change.

---

### Notes:

- Noticed a memory regression described [here](https://github.com/tenstorrent/tt-blacksmith/issues/625).
- The LLaMA 70B numbers look a bit suspicious. Will look into during trace effort. https://github.com/tenstorrent/tt-blacksmith/issues/626
- AOT Autograd will require more work on this since by default it joins them which is undesirable since tensors remain persistent. There are a few ways to tackle this, currently prototyping. 

---

### What's Changed
- Change LLaMA script to include performance improved calls.
- Make model compilation configurable.

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 1]
```diff
@@ -135,13 +130,17 @@ def train(
 
     tokenizer = train_dataset.tokenizer
 
+    compile_options = {"tt_enable_torch_fx_fusion_pass": False, "tt_legacy_compile": True, "tt_lazy_execution": True}
```
Comment (ndrakulicTT) [inline blacksmith/experiments/torch/llama/xla/train.py]: Why not enable `tt_lazy_execution` by default in `blacksmith/models/torch/huggingface/hf_models.py`

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 2]
```diff
@@ -184,11 +183,15 @@ def train(
...
+                loss_ = train_fn(batch, model, cross_entropy_loss, config.gradient_accumulation_steps)
+                loss_.backward()
 
-                if config.use_tt:
-                    torch_xla.sync(wait=True)
+                # Sync only gradients and loss.
+                tensors_to_sync = [loss_] + [p.grad for p in trainable_params if p.grad is not None]
+                devices = list({t.device.type for t in tensors_to_sync})
+                torch_xla._XLAC._xla_sync_multi(tensors_to_sync, devices, wait=True)
+                torch_xla._XLAC._clear_pending_irs(torch_xla._XLAC._xla_get_default_device())
```
Comment (ndrakulicTT) [inline blacksmith/experiments/torch/llama/xla/train.py]: For my info, what does `_clear_pending_irs` do?

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 3]
```diff
@@ -184,11 +183,15 @@ def train(
...
 
                 running_loss += loss_.item()
```
Comment (abogdanovicTT) [inline blacksmith/experiments/torch/llama/xla/train.py:202]: We return `scaled_loss` and add that to `running_loss`, does it mean `avg_loss` will be smaller?

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 1 above)
Comment (agobeljicTT) [reply blacksmith/experiments/torch/llama/xla/train.py]: maybe localize the change first?

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 4]
```diff
@@ -87,20 +87,14 @@ def validate(model, val_data_loader, loss_fn, logger, device, config, tokenizer=
...
 def training_step_inner(batch, model, loss_fn, gradient_accumulation_steps):
     output = model(input_ids=batch["input_ids"], attention_mask=batch["attention_mask"])
     logits = output.logits
     shift_logits = logits[:, :-1, :].contiguous()
     loss = loss_fn(shift_logits, batch["expected_output"], batch["labels_mask"])
     # Scale loss by number of accumulation steps to get correct effective batch size.
     scaled_loss = loss / gradient_accumulation_steps
-    scaled_loss.backward()
-    return loss.detach()
+    return scaled_loss
```
Comment (agobeljicTT) [inline blacksmith/experiments/torch/llama/xla/train.py]: this function now doesn't perform backward, making its name a bit incorrect
perhaps something like forward pass

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 2 above)
Comment (agobeljicTT) [inline blacksmith/experiments/torch/llama/xla/train.py]: guard with config.use_tt

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 3 above)
Comment (agobeljicTT) [reply blacksmith/experiments/torch/llama/xla/train.py:202]: yep, this should be changed

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 4 above)
Comment (pglusacTT) [reply blacksmith/experiments/torch/llama/xla/train.py]: Good point!

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 1 above)
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/train.py]: This will break GPU run, since we are using "tt" backend.

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 5]
```diff
@@ -135,13 +130,17 @@ def train(
 
     tokenizer = train_dataset.tokenizer
 
+    compile_options = {"tt_enable_torch_fx_fusion_pass": False, "tt_legacy_compile": True, "tt_lazy_execution": True}
+    train_fn = torch.compile(training_step_inner, backend="tt", options=compile_options)
+    model_compiled = torch.compile(model, backend="tt", options=compile_options)
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/train.py]: I am not a fan of this going here without a guard. If someone flips `compile_model` in `get_model` we will error out here.

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 3 above)
Comment (pglusacTT) [reply blacksmith/experiments/torch/llama/xla/train.py:202]: Yes, good catch

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 1 above)
Comment (pglusacTT) [reply blacksmith/experiments/torch/llama/xla/train.py]: Agreed with @agobeljicTT . It is out of scope for the effort but I'd push for it in https://github.com/tenstorrent/tt-blacksmith/issues/615.

Comment (mmilosevicTT) [reply blacksmith/experiments/torch/llama/xla/train.py]: As we discussed offline. I agree with @ndrakulicTT , but I would just check that pushing compilation of model to `get_model` works without this refactor. We can always use different flag `lazy_execution` for options where we want to enable this flag and not keep it that full compilation goes out of `get_model`.

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 2 above)
Comment (pglusacTT) [reply blacksmith/experiments/torch/llama/xla/train.py]: There was redundant recomputation of gradients done in the optimizer IR (still not 100% sure why, I think it is related to some gradient view that is not synced). After backward, optimizer will do a `mark_step` which will compute any tensors that are left uncomputed.  With `_clear_pending_irs` you effectively just disregard any compute that isn't related to the optimizer (like view materialization etc). 

I double checked the grad values and they are the same in the old path and in the new one, the training curve follows expected values. Tried accumulation as well, everything is ok.

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 6]
```diff
@@ -135,13 +127,25 @@ def train(
...
     tokenizer = train_dataset.tokenizer
 
+    if config.use_tt:
+        compile_options = {
+            "tt_enable_torch_fx_fusion_pass": False,
+            "tt_legacy_compile": True,
+            "tt_lazy_execution": True,
+        }
+        compute_loss_fn = torch.compile(compute_loss, backend="tt", options=compile_options)
+        eval_model = torch.compile(model, backend="tt", options=compile_options)
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/train.py]: Just out of curiosity, I see you noticed DRAM regression, but have you checked the host memory usages for this, as well? Does torch use same underlying model for both compilations or does it create a copy in memory? 

I am worried for this as we already have opened issues for TT -> CPU movements of models where memory leaks constantly in each loop (https://github.com/tenstorrent/tt-xla/issues/5292).

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 7]
```diff
@@ -184,11 +188,17 @@ def train(
                 # Shard model if tensor parallelism is used.
                 device_manager.shard_model(model)
 
-                # Training step.
-                loss_ = training_step_inner(batch, model, cross_entropy_loss, config.gradient_accumulation_steps)
+                loss_ = compute_loss_fn(batch, model, cross_entropy_loss)
+                # Scale by the accumulation steps so the effective-batch gradient is the mean.
+                scaled_loss = loss_ / config.gradient_accumulation_steps
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/train.py]: Why don't we leave this in `compute_loss_fn` and just remove scaling in logging and running loss accumulation?

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 8]
```diff
@@ -184,11 +188,17 @@ def train(
...
+                scaled_loss.backward()
 
                 if config.use_tt:
-                    torch_xla.sync(wait=True)
+                    tensors_to_sync = [loss_] + [p.grad for p in trainable_params if p.grad is not None]
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/train.py:196]: Should't we use `scaled_loss` here?

File: blacksmith/experiments/torch/llama/xla/train.py [hunk 9]
```diff
@@ -135,13 +127,25 @@ def train(
...
+    if config.use_tt:
+        compile_options = {
+            "tt_enable_torch_fx_fusion_pass": False,
+            "tt_legacy_compile": True,
+            "tt_lazy_execution": True,
+        }
+        compute_loss_fn = torch.compile(compute_loss, backend="tt", options=compile_options)
+        eval_model = torch.compile(model, backend="tt", options=compile_options)
+    else:
+        compute_loss_fn = compute_loss
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/xla/train.py]: nit: Not a fan of this naming `compute_loss_fn`. Generally compute is redundant here if you ask me, `loss_fn` itself implies calculation.

Comment (mmilosevicTT) [reply blacksmith/experiments/torch/llama/xla/train.py]: Maybe even rename `compute_loss` to something like `scoped_loss_fn` or even `unscaled_loss_fn` if you decide not to move scaling to the loop.

**Review by mmilosevicTT — COMMENTED** [review]
Looks good in general.

I only have 2 concerns in line 137 and 197. Otherwise nits 😄

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 9 above)
Comment (pglusacTT) [reply blacksmith/experiments/torch/llama/xla/train.py]: Renamed: `compute_loss_fn` -> `loss_step` and `compute_loss` -> `scoped_loss_fn` (kept the `scoped` part since the function still scopes the large logits tensor locally).

File: blacksmith/experiments/torch/llama/xla/train.py (hunk 6 above)
Comment (pglusacTT) [reply blacksmith/experiments/torch/llama/xla/train.py]: Good question. `torch.compile` does not copy the model; both `loss_step` (which receives `model` as an argument) and `eval_model = torch.compile(model)` reference the same `nn.Module` instance and the same parameter tensors (`eval_model._orig_mod is model`), so there is no duplicate set of weights on host. The two compiles only produce two cached graphs/executables (forward-only for validation vs. forward+loss+backward for training), which costs some host memory for the compiled artifacts but not a second copy of the weights.

**Review by mmilosevicTT — APPROVED** [review]
LGTM

nit: As we discussed offline, I realise that my previous comment is a bit off. `loss_fn` should regard to just logits and loss. Could you revert to `compute_loss`.
